src/ocean_emulators/utils/loss.py

```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
     mae = F.l1_loss(pred, target, reduction="none")
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
+
+
+N_WINDOW = 25
+
+
+class MseDynamic:
+    """A loss function that scales each channel contribute equally to the loss.
+
+    See: https://m2lines.slack.com/files/U086XUETAJ0/F0952H4G3GF/dynamic_loss_weighting.pptx
+    """
+
+    def __init__(
+        self,
+        wet: Grid,
+        stds: Float[torch.Tensor, " var"],
+        *,
+        should_limit: bool,
+    ):
+        self.wet: Grid = wet
+        self.per_channel_scale: Float[torch.Tensor, " var"] = torch.ones(
+            stds.shape[0], device=wet.device
+        )
+        if should_limit:
+            vars: Float[torch.Tensor, " var"] = stds.pow(2)
+            self.limits: Float[torch.Tensor, " var"] | None = 1.0 / vars
+        else:
+            self.limits = None
+
+    def __call__(
+        self,
+        pred: Float[torch.Tensor, "batch hist*var lat lon"],
+        target: Float[torch.Tensor, "batch hist*var lat lon"],
+    ) -> Float[torch.Tensor, " hist*var"]:
+        loss_with_history_channels: Float[torch.Tensor, " hist*var"] = decomposed_mse(
+            pred, target, self.wet
+        )
+        scaled_loss_including_history_dimension: Float[torch.Tensor, "hist var"] = (
+            loss_with_history_channels.reshape(self.per_channel_scale.shape[0], -1)
+            * self.per_channel_scale.unsqueeze(1)
+        )
+        return scaled_loss_including_history_dimension.reshape(-1)
+
+    def update(
+        self,
+        pred: Float[torch.Tensor, "batch hist*var lat lon"],
+        target: Float[torch.Tensor, "batch hist*var lat lon"],
+    ):
+        mse_loss = decomposed_mse(pred, target, self.wet)
+        mse_loss = torch.where(mse_loss == 0, 1e-8, mse_loss)
+        new_target_weights_with_history: Float[torch.Tensor, " hist*var"] = (
+            1.0 / mse_loss
+        )
+        # Reshape from channels * history to channels by averaging
```

Comment:
:sheep: I was a little confused until I parsed this more in detail

```suggestion
        # Reshape from channels * history to channels by averaging (along the `hist` dimension)
```